Update Azure API Guidelines for new LRO pattern (#353)
Microsoft/api-guidelines: azure/ConsiderationsForServiceDesign.md

# Considerations for Service Design

<!-- cspell:ignore autorest, etag, idempotency -->

## History

| Date        | Notes                                                          |
| ----------- | -------------------------------------------------------------- |

> | 2022-May-20 | Update guidance on long-running operations                     |

Reviewer: _(Nit - We've rolled over to June already)_

operation request.

This diagram illustrates how a long-running operation with a status monitor is initiated and then how the client
determines it has completed and obtains its results:

```mermaid
sequenceDiagram
    participant Client
    participant API Endpoint
    participant Status Monitor
    Client->>API Endpoint: PUT/POST/DELETE
    API Endpoint->>Client: HTTP/1.1 202 Accepted<br/>Retry-After: 5<br/>{ "id": "22", "status": "NotStarted" }
    Client->>Status Monitor: GET
    Status Monitor->>Client: HTTP/1.1 200 Ok<br/>Retry-After: 5<br/>{ "id": "22", "status": "Running" }
    Client->>Status Monitor: GET
    Status Monitor->>Client: HTTP/1.1 200 Ok<br/>Retry-After: 5<br/>{ "id": "22", "status": "Succeeded" }
```

1. The client sends the request to initiate the long-running operation.
The initial request could be a PUT, POST, or DELETE method.
PATCH must never be used for LROs -- it should be reserved for simple resource updates.
If a long-running update is required it should be implemented with POST.
The request may contain an `operation-id` header that the service uses as the ID of the status monitor created for the operation.

> 2. The service validates the request and initiates the operation processing.

Reviewer 2: Should we say something about what happens if the request is invalid? For example..
> If the request is not valid, the service will return a `400` level error, ideally with the reason the request was invalid. It is important to realize that this occurs prior to the LRO starting, and is therefore, NOT a failure of the long running operation.
I might be over thinking this, but for folks that are not familiar LROs, it might be good to ensure they realize that the incoming request could fail.

before polling (GETting) the status monitor URL again for an update.
For backward compatibility, the response may also include an `Operation-Location` header containing the absolute URL
of the status monitor resource (without an api-version query parameter).

3. After waiting at least the amount of time specified by the previous response's `Retry-after` header,
the client issues a GET request to the status monitor using the ID in the body of the initial response.
The GET operation for the status monitor is documented in the REST API definition and the ID
is the last URL path segment.

4. The status monitor responds with information about the operation including its current status,
which should be represented as one of a fixed set of string values in a field named `status`.
If the operation is still being processed, the status field will contain a "non-terminal" value, like `NotStarted` or `Running`.

5. After the operation processing completes, a GET request to the status monitor returns the status monitor with a status field set to a terminal value -- `Succeeded`, `Failed`, or `Canceled` -- that indicates the result of the operation.

> If the status is `Failed` or `Canceled`, the status monitor resource contains an `error` field with a `code` and `message` that describes the failure.

Reviewer 2: Is `Canceled` an error? I think it is not. IMHO, `Canceled` is the result of a deliberate action, either by a user or another system. Because it is a deliberate action, it's not an error. Could we treat this as another terminal state but include a string for "reason" ? 

Since canceling an LRO is a POST, we could potentially pass in a reason in the body, e.g. 
POST /<status-monitor-url>:cancel
{ "reason":"i got tired of waiting"}

Reviewer 3: The LRO "failed" to complete due to cancellation. So, since the LRO is not successful, I think the error should indicate that it failed due to cancellation.

Author: I discussed with Jeffrey today and he is willing to reconsider, so we'll drop cancelled for now and see if we get feedback on this.

Service may offer DELETE due to GDPR/privacy.

### Long-running Action Operations

An action operation that is also long-running combines the [Action Operations](#action-operations) pattern 
with the [Long Running Operations](#long-running-operations) pattern.

The operation is initiated with a POST operation and the operation path ends in `:action`.

```text
POST /<service-or-resource-url>:action 
Operation-Id: 22 
 
{ 
   "arg1": 123 
   "arg2": "abc" 
} 
```

The response is a `202 Accepted` as described above.

```text
HTTP/1.1 202 Accepted
Operation-Location: https://<status-monitor-endpoint>/22
Retry-After: 5 
 
{
   "id": "22",
   "status": "NotStarted"
}
```

The client will issue a GET to the status monitor to obtain the status and result of the operation.

```text
GET https://<status-monitor-endpoint>/22?api-version=2022-05-01
```

When the operation completes successfully, the result (if there is one) will be included in the `result` field of the status monitor.

```text
HTTP/1.1 200 Ok
 
{
   "id": "22",
   "status": "Succeeded",
   "result": { ... }
}
```

### Create (PUT) with additional long-running processing

> A special case of long-running operation that occurs often is a PUT operation to create a resource

Reviewer 2: We should give an example of this to make this a bit more clear.

Author: Added.

Long-running operations are an API design pattern that should be used when the processing of
an operation may take a significant amount of time -- longer than a client will want to block
waiting for the result.
The request that initiates a long-running operation generally returns a `202 Accepted` response with a
response body that contains a representation of a _status monitor_, which is an ephemeral resource that
will track the status and final result of the operation.

> The status monitor resource is distinct from the target resource and specific to the individual

Reviewer 3: ```suggestion
The status monitor resource is distinct from the target resource (if any) and specific to the individual
```

operation request.

This diagram illustrates how a long-running operation with a status monitor is initiated and then how the client
determines it has completed and obtains its results:

```mermaid
sequenceDiagram
    participant Client
    participant API Endpoint
    participant Status Monitor
    Client->>API Endpoint: PUT/POST/DELETE
    API Endpoint->>Client: HTTP/1.1 202 Accepted<br/>Retry-After: 5<br/>{ "id": "22", "status": "NotStarted" }
    Client->>Status Monitor: GET
    Status Monitor->>Client: HTTP/1.1 200 Ok<br/>Retry-After: 5<br/>{ "id": "22", "status": "Running" }
    Client->>Status Monitor: GET
    Status Monitor->>Client: HTTP/1.1 200 Ok<br/>Retry-After: 5<br/>{ "id": "22", "status": "Succeeded" }
```

1. The client sends the request to initiate the long-running operation.
The initial request could be a PUT, POST, or DELETE method.

> PATCH should not be used for LROs -- it should be reserved for simple resource updates.

Reviewer 3: ```suggestion
PATCH must never be used for LROs -- it should be reserved for simple resource updates.
```

If a long-running update is required it should be implemented with POST.

> The request may contain an `operation-id` header that the service will use as the ID of the status monitor created for the operation.

Reviewer 3: ```suggestion
The request may contain an `operation-id` header that the service uses as the ID of the status monitor created for the operation.
```

> 2. The resource validates the request and initiates the operation processing.

Reviewer 3: ```suggestion
2. The service validates the request and initiates the operation processing.
```

It sends a response to the client with a `202-Accepted` HTTP status code.

> The response body is status monitor for the operation including the ID, either from the request header or generated by the service.

Reviewer 3: ```suggestion
The response body is the status monitor for the operation including the ID, either from the request header or generated by the service.
```

> The response also includes a `Retry-after` header telling the client a minimum time to wait (in seconds)

Reviewer 3: ```suggestion
When returning a status monitor whose status is not in a terminal state, the response must also include a `retry-after` header indicating the minimum number of seconds the client should wait 
```

> before sending a request to the status monitor URL.

Reviewer 3: ```suggestion
before polling (GETting) the status monitor URL again for an update.
```

> For backward compatibility, the response may also include an `Operation-Location` header containing the absolute URL

Reviewer 3: may or must? I think we want must, right?

Author: I said "may" because Operation-Location is redundant with the information in the status monitor and thus not really necessary.  Our current SDK patterns use it so _for backward compatibility_ it is good to have but hopefully we'll update the generators soon so that they operate on the status monitor information and not the `Operation-Location`.

Reviewer 4: I would strongly push for MUST include the Operation-Location. The operation location is not redundant in the response for anything that wants to act in a hyperlinked way. The status monitor resource by id only requires out-of-band knowledge in order to correctly form the URL. And while hyperlinking is not super commonly used in the general case, for single redirects etc. it has certainly proven to be a valuable concept in HTTP.

of the status monitor resource (without an api-version query parameter).

3. After waiting at least the amount of time specified by the previous response's `Retry-after` header,
the client issues a GET request to the status monitor using the ID in the body of the initial response.
The GET operation for the status monitor is documented in the REST API definition and the ID

> is a path parameter.

Reviewer 3: ```suggestion
is the last URL path segment.
```

4. The status monitor responds with information about the operation including its current status,
which should be represented as one of a fixed set of string values in a field named `status`.
If the operation is still being processed, the status field will contain a "non-terminal" value, like `NotStarted` or `Running`.

5. After the operation processing completes, a GET request to the status monitor returns the status monitor with a status field set to a terminal value -- `Succeeded`, `Failed`, or `Canceled` -- that indicates the result of the operation.
If the status is `Failed` or `Canceled`, the status monitor resource contains an `error` field with a `code` and `message` that describes the failure.
If the status is `Succeeded` and the LRO is an Action operation, the operation results will be returned in the `results` field of the status monitor.

> If the status is `Succeeded` and the LRO is a CRUD operation on a resource, the client can perform a GET on the resource

Reviewer 3: CRUD is a bit much; it can only be create, update or delete; not read.

Author: Agreed.  I think "CRUD" is unnecessary here.
```suggestion
If the status is `Succeeded` and the LRO is an operation on a resource, the client can perform a GET on the resource
```

to observe the result of the operation if desired.

6. There may be some cases where a long-running operation can be completed before the response to the initial request.

> In these cases, the operation should still return a `202 Accepted` but the `status` in the response body be set to the

Reviewer 3: ```suggestion
In these cases, the operation should still return a `202 Accepted` with the `status` property set to the appropriate terminal state.
```

> terminal value `Succeeded` so that the client does not need to make an additional network request.

Reviewer 3: I'd delete this. The operation can fail synchronously too.

Author: Agreed.
```suggestion
```

Long-running operations are an API design pattern that should be used when the processing of
an operation may take a significant amount of time -- longer than a client will want to block
waiting for the result.

The request that initiates a long-running operation generally returns a `202 Accepted` response with a

> response body that contains a representation of a _status monitor_, which is an ephemeral resource that

Reviewer 4: The 202 + status monitor response body should only be returned for a POST and DELETE operation. So I'm not sure if the "generally" vs. "special case" terminology is the right choice. 2 vs. 1 may not be a 100% fair fight, but it also doesn't reach my "generally" bar :). 

I think we are also falling into the same trap that the old guidance did in that in order to correctly understand the guidance, you have to understand HTTP terminology. While it is true that the status monitor is a resource (in HTTP terminology), my question is how likely it is that our target audience will correctly interpret, or if their understanding of a "resource" is different. I've heard "it's not a resource, it is just a URL or location" plenty of times in my discussions with service teams.

Author: Now called "POST or DELETE LRO".

will track the status and final result of the operation.
The status monitor resource is distinct from the target resource (if any) and specific to the individual
operation request.
There is a special form of long-running operation initiated with PUT and returns `201 Created` that is described
in [Create (PUT) with additional long-running processing](#create-put-with-additional-long-running-processing).
The remainder of this section applies to all long-running operations _except_ for the create with PUT type LRO.

This diagram illustrates how a long-running operation with a status monitor is initiated and then how the client
determines it has completed and obtains its results:

```mermaid
sequenceDiagram
    participant Client
    participant API Endpoint
    participant Status Monitor

>     Client->>API Endpoint: PUT/POST/DELETE

Reviewer 4: Per above, PUT should not respond with a 202 .

which should be represented as one of a fixed set of string values in a field named `status`.
If the operation is still being processed, the status field will contain a "non-terminal" value, like `NotStarted` or `Running`.

5. After the operation processing completes, a GET request to the status monitor returns the status monitor with a status field set to a terminal value -- `Succeeded`, `Failed`, or `Canceled` -- that indicates the result of the operation.
If the status is `Failed`, the status monitor resource contains an `error` field with a `code` and `message` that describes the failure.
If the status is `Succeeded` and the LRO is an Action operation, the operation results will be returned in the `results` field of the status monitor.
If the status is `Succeeded` and the LRO is an operation on a resource, the client can perform a GET on the resource
to observe the result of the operation if desired.

6. There may be some cases where a long-running operation can be completed before the response to the initial request.
In these cases, the operation should still return a `202 Accepted` with the `status` property set to the appropriate terminal state.

> 7. The service will auto-purge status monitor resource after completion (at least 24 hours).

Reviewer 4: This sentence is broken.

Long-running operations are an API design pattern that should be used when the processing of
an operation may take a significant amount of time -- longer than a client will want to block
waiting for the result.

The request that initiates a long-running operation returns a response that points to or embeds
a _status monitor_, which is an ephemeral resource that will track the status and final result of the operation.
The status monitor resource is distinct from the target resource (if any) and specific to the individual
operation request.

A POST or DELETE operation returns a `202 Accepted` response with the status monitor in the response body.
A long-running POST should not be used for resource create -- use PUT as described below.

> PATCH must never be used for LROs -- it should be reserved for simple resource updates.

Reviewer 5: nit, before we use `LRO`, we probably want to make sure reader knows that `LROs == long-running operation`

> PATCH must never be used for LROs -- it should be reserved for simple resource updates.

Author: ```suggestion
PATCH must never be used for long-running operations -- it should be reserved for simple resource updates.
```

If a long-running update is required it should be implemented with POST.

There is a special form of long-running operation initiated with PUT that is described
in [Create (PUT) with additional long-running processing](#create-put-with-additional-long-running-processing).
The remainder of this section describes the pattern for long-running POST and DELETE operations.

This diagram illustrates how a long-running operation with a status monitor is initiated and then how the client
determines it has completed and obtains its results:

```mermaid
sequenceDiagram
    participant Client
    participant API Endpoint
    participant Status Monitor
    Client->>API Endpoint: POST/DELETE
    API Endpoint->>Client: HTTP/1.1 202 Accepted<br/>Retry-After: 5<br/>{ "id": "22", "status": "NotStarted" }
    Client->>Status Monitor: GET

>     Status Monitor->>Client: HTTP/1.1 200 Ok<br/>Retry-After: 5<br/>{ "id": "22", "status": "Running" }

Reviewer 6: Nit: All `Ok` -> `OK`.

    Client->>Status Monitor: GET
    Status Monitor->>Client: HTTP/1.1 200 Ok<br/>{ "id": "22", "status": "Succeeded" }
```

1. The client sends the request to initiate the long-running operation.

> The initial request could be a PUT, POST, or DELETE method.

Reviewer 5: It is a bit strange that
`The remainder of this section describes the pattern for long-running POST and DELETE operations.`
then a few lines later
`The initial request could be a PUT, POST, or DELETE method.`

> The initial request could be a PUT, POST, or DELETE method.

Author: ```suggestion
The initial request could be a POST or DELETE method.
```

> The request may contain an `operation-id` header that the service uses as the ID of the status monitor created for the operation.

Reviewer 5: Should we specify what is a valid `operation-id`? I guess we don't want customer to inject an URL to service (to the status monitor URL). (SDK likely know to use a GUID)

Is it "MUST" in POST? If we lost the response, without it, we seems not able to get the result.
If customer uses REST API, they had to know how to handle the status monitor. So it might not be too much to ask them always provides the `operation-id` in request?

Author: We have debated the optional vs required for `operation-id` and (maybe grudgingly) decided on optional. We could reopen the discussion ...

Reviewer 5: No, we don't need to reopen the discussion (yes, remember discussion on whether to merge it with `Repeatability-Request-ID`).

What I think to be important is the specification on what value of `operation-id` is acceptable (e.g. GUID only?). This way, both user and backend developer would be clear on what to send and what to expect/validate (I assume user cannot send `!@#$%^&*(` as `operation-id`), and we prefer it to be consistent across services.
I know this is not an RFC, but still want it to be precise.

Reviewer 7: Perhaps an example of what an operation-id can be (whether GUID) and how it's used from the clients perspective and what are the implications of it not being present.

Reviewer 7: I see there is some wording around an operation-id that it can be a GUID in the Guidelines.md file

Author: When we update the Azure API Style guide I think we can say that the operation-id header should document allowed characters (with `pattern`) and maximum length, which is already documented for ids passed in path parameters.

https://github.com/Azure/azure-api-style-guide/blob/main/openapi-style-guide.md#path-parameters

Reviewer 5: OK.

Though I would suggest that we say here that GUID "must" be supported, whatever other allowed pattern.
This would give user the minimal guarantee that if they use a GUID, it will be accepted, whatever service one is communicating (another advantage is that user be able to use the same value of `Repeatability-Request-ID` as `operation-id`).
I know we probably not having lots of LRO in data-plane, but if user uses data-plane REST API from multiple services, we could save them the time to read doc on every service to decide which value can be used (there is the difference with resource ID, which user probably had to read the doc on each service they use).

Reviewer 4: I think anything other than using a GUID is inviting trouble since most well behaved services would use a GUID (or possibly some form of hash for the request - but see below) in order to avoid name conflicts (which are really bad for operations). The GUID-to-string conversion is sensitive to a couple of different things; with-or-without curly braces, with-or-without dashes and case sensitivity.

Reviewer 7: > 

I believe there are two aspects here:

1) The 'operation-id' being recommended and the edge cases covered here. 
2) The 'operation-id' being a type that's in-depth covered in the style guideline with recommendations around the normalization and code generation

2. The service validates the request and initiates the operation processing.
If there are any problems with the request, the service responds with a `4xx` status code and error response body.
Otherwise the service responds with a `202-Accepted` HTTP status code.

> The response body is the status monitor for the operation including the ID, either from the request header or generated by the service.

Reviewer 5: Should we say that the response "MUST" contains an `id` property (if 202)?

Is `status` also a "MUST" in response of initial request?

Reviewer 5: I am not sure if it is just me, but `id` in response of initial request of POST could be confusing with the same `id` in response of initial request of PUT. The former is the ID for status monitor, the latter is the ID for the resource.

Author: POST should never be used to create resources and especially not an LRO POST, so there should be no confusion about the `id` in the response body of LRO POST.

LRO PUT will return the resource in the body and the operation ID in the `Operation-Id` header, so I think this is also clear which ID is for which.

I think this is enough to avoid confusion for most users, but you want to suggest something that would make it even clearer please feel free.

Reviewer 6: I wouldn't say *never*. Consider a Key Vault certificate. The request to create is significantly different from the resource created just by the very nature of CSRs vs. certificates. From meetings, I thought we agree this is rare and maybe doesn't need to be detailed, but shouldn't we at least acknowledge it is a (less desirable) possibility lest we have different teams coming up with different solutions?

Reviewer 6: ...in fact, it's documented below and acknowledged by Jeffrey.

Reviewer 5: Appears that I am the only one confused about the `id`. Let's drop it.

---

The POST to create resource would be a different topic.

I assume if we do that, the (created) resource ID (and maybe the full URL of the resource) would be provided in the `result` property, in the success case of status monitor poll (I could be wrong, as we didn't get to the discussion on this scenario in detail).
```
HTTP/1.1 200 Ok

{
   "id": "22",
   "status": "Succeeded",
   "result": { ... }
}
```

Author: I added a statement in Guidelines.md:

> :no_entry: **DO NOT** use a long-running POST to create a resource -- use PUT as described below.

If some service insists on using POST rather than PUT we'll probably guide them into a design that looks like the PUT pattern, but we shouldn't offer this as an option in the Guidelines.

Reviewer 4: There are valid uses cases for POST for resource creation. It should be avoided, yes, and certainly not used for service implementation convenience, but for use cases where it is imperative that the service owns the uniqueness of the identity of a resource (possibly global uniqueness), POST would be the preferred pattern. I don't mind this verbiage in the guidelines as long as we, as a stewardship board, understand when it is and when it is not appropriate to use POST for resource creates.

When returning a status monitor whose status is not in a terminal state, the response must also include a `retry-after` header indicating the minimum number of seconds the client should wait

> before polling (GETting) the status monitor URL again for an update.

Reviewer 5: ```suggestion
before polling (GETing) the status monitor URL again for an update.
```

which should be represented as one of a fixed set of string values in a field named `status`.
If the operation is still being processed, the status field will contain a "non-terminal" value, like `NotStarted` or `Running`.

5. After the operation processing completes, a GET request to the status monitor returns the status monitor with a status field set to a terminal value -- `Succeeded`, `Failed`, or `Canceled` -- that indicates the result of the operation.
If the status is `Failed`, the status monitor resource contains an `error` field with a `code` and `message` that describes the failure.
If the status is `Succeeded` and the LRO is an Action operation, the operation results will be returned in the `results` field of the status monitor.
If the status is `Succeeded` and the LRO is an operation on a resource, the client can perform a GET on the resource
to observe the result of the operation if desired.

6. There may be some cases where a long-running operation can be completed before the response to the initial request.
In these cases, the operation should still return a `202 Accepted` with the `status` property set to the appropriate terminal state.

> 7. The service will auto-purge the status monitor resource after completion (at least 24 hours).

Reviewer 7: Perhaps we should include a short paragraph with examples illustrating the edge cases/negative cases. Ex: client fails to receive the response for LRO and retries with the same ID or when the ID is missing, etc.. What are the other interesting edge cases?

Reviewer 7: The case with the same ID is covered in Guidelines.md but I wonder if the reader would benefit from a paragraph of examples with edge cases

Reviewer 2: @MushMal -- How different are the examples you mentioned from "normal" REST calls? For example, if I fail to get a response from a "regular" call, wouldn't this work the same way? I'm thinking we can simply point to other parts of the guidelines and "demystify" LROs a bit more, e.g. "they are just like other REST calls"...

Author: We'll watch for patterns in edge cases in API reviews and feed those back into these guidelines.

The service may offer DELETE of the status monitor resource due to GDPR/privacy.

### Long-running Action Operations

An action operation that is also long-running combines the [Action Operations](#action-operations) pattern
with the [Long Running Operations](#long-running-operations) pattern.

The operation is initiated with a POST operation and the operation path ends in `:action`.

```text

> POST /<service-or-resource-url>:action 

Reviewer 6: Should include an api-version like we do for other examples.

Operation-Id: 22 
 
{ 
   "arg1": 123 
   "arg2": "abc" 
} 
```

The response is a `202 Accepted` as described above.

```text
HTTP/1.1 202 Accepted
Operation-Location: https://<status-monitor-endpoint>/22
Retry-After: 5 
 
{
   "id": "22",
   "status": "NotStarted"
}
```

The client will issue a GET to the status monitor to obtain the status and result of the operation.

```text
GET https://<status-monitor-endpoint>/22?api-version=2022-05-01
```

When the operation completes successfully, the result (if there is one) will be included in the `result` field of the status monitor.

```text
HTTP/1.1 200 Ok
 
{
   "id": "22",
   "status": "Succeeded",
   "result": { ... }
}
```

> ### Create (PUT) with additional long-running processing

Reviewer 8: I might be missing something. But in practice, PUT is often used to either create and update. Is there a different behavior  for updating existing resources with a long-running operation? I feel the guideline can be more prescriptive on it and looks like it should be consistent with create case, i.e. updated content is applied first before LRO is finished.

Reviewer 3: We do NOT consider PUT to create or update; we consider PUT to be create or wholesale replace. We consider PATCH to be create or update. The reason for this is due to Azure's versioning policies. I do have a video describing this [here](https://www.youtube.com/watch?v=PFK6RRTwXCg).

Reviewer 8: Jeffrey, "update" might be a bit misleading but it is used in many Azure API docs for PUT. One example is "[Create Or Update](https://docs.microsoft.com/en-us/rest/api/compute/virtual-machines/create-or-update)" . But my point is to make the guideline specific on the behavior of PUT in the case of replacement. I would expect it is the same as create case except response code would be 202 instead of 201.

Reviewer 3: To update a resource via an LRO, we prescribe the POST-action LRO pattern.

Author: I changed the wording a bit to make clear that the "PUT" pattern applies to both "create" and "replace" (since they really aren't separable).

> ### Create (PUT) with additional long-running processing

Author: ```suggestion
### PUT with additional long-running processing
```

Reviewer 6: If you make this change, don't forget to change the link text and bookmark on or around line 159.

> A special case of long-running operation that occurs often is a PUT operation to create a resource

Author: ```suggestion
A special case of long-running operation that occurs often is a PUT operation to create or replace a resource
```

that involves some additional long-running processing.
One example is a resource requires physical resources (e.g. servers) to be "provisioned" to make the resource functional.
In this case, the request may contain an `operation-id` header that the service will use as
the ID of the status monitor created for the operation.

```text
PUT /items/FooBar&api-version=2022-05-01
Operation-Id: 22

{
   "prop1": 555, 
   "prop2": "something"
}
```

> In this case the response to the initial request is a `201 Created` to indicate that the resource has been created.

Author: ```suggestion
In this case the response to the initial request is a `201 Created` to indicate that the resource has been created, or `200 OK` when the resource was replaced.
```

The response body contains a representation of the created resource, which is the standard pattern for a create operation.
A status monitor is created to track the additional processing and the ID of the status monitor
is returned in the `Operation-Id` header of the response.
The response may also include an `Operation-Location` header for backward compatibility.
If the resource supports ETags, the response may contain an `etag` header and possibly an `etag` property in the resource.

```text
HTTP/1.1 201 Created 
Operation-Id: 22
Operation-Location: https://items/operations/22
etag: "123abc"

{
  "id": "FooBar", 
  "etag": "123abc",
  "prop1": 555,
  "prop2": "something"
}
```

The client will issue a GET to the status monitor to obtain the status of the operation performing the additional processing.

```text
GET https://items/operations/22?api-version=2022-05-01
```

When the additional processing completes, the status monitor will indicate if it succeeded or failed.

```text
HTTP/1.1 200 Ok
 
{
   "id": "22",
   "status": "Succeeded"
}
```

> If the additional processing failed, the service may delete the original resource if it is not usable in this state.

Reviewer 5: the "MAY" here might be ambiguous for customer. Seems the only option for them would be always manually check and delete, if service "may" delete the resource.

Author: I think the service would need to clearly document this behavior. I suppose we could make that explicit here but I feel that that's generally implied across all topics, so adding it here might mean we have to be explicit everywhere else too.

### Long-running delete operation

A long-running delete operation follows the general pattern of a long-running operation --
it returns a `202 Accepted` with a status monitor which the client uses to determine the outcome of the delete.

The resource being deleted should remain visible (returned from a GET) until the delete operation completes successfully.

When the delete operation completes successfully, a client must be able to create new resource with same name without conflicts.

> 

Reviewer 7: Perhaps we should mention what would happen if a client issues a POST/PATCH while the LRO delete operation is ongoing and the resource is still visible. This would be good for the implementation.

### Controlling a long-running operation

It might be necessary to support some control action on a long-running operation, such as cancel.
This is implemented as a POST on the status monitor endpoint with `:action` added.

```text

> POST /<status-monitor-url>:cancel

Reviewer 6: Needs api-version (consistency).

```

> A successful response to a control operation should be a `200 Ok` with a representation of the status monitor.

Reviewer 6: Do we want to prescribe whether this is the same status monitor as the original being changed? For example, if you already have a separate or embedded status monitor, canceling might return the same one and the status monitor property e.g., `provisioningState` would just eventually change to "canceled". Or is that really more of an implementation detail we needn't worry about?

Author: Our guidance will be to avoid long-running cancel and we'll see if that works.

Long-running operations are an API design pattern that should be used when the processing of
an operation may take a significant amount of time -- longer than a client will want to block
waiting for the result.

The request that initiates a long-running operation returns a response that points to or embeds
a _status monitor_, which is an ephemeral resource that will track the status and final result of the operation.
The status monitor resource is distinct from the target resource (if any) and specific to the individual
operation request.

A POST or DELETE operation returns a `202 Accepted` response with the status monitor in the response body.
A long-running POST should not be used for resource create -- use PUT as described below.
PATCH must never be used for long-running operations -- it should be reserved for simple resource updates.
If a long-running update is required it should be implemented with POST.

> There is a special form of long-running operation initiated with PUT that is described

Reviewer 2: This sentence and the next are distracting at this point in the document. Because this is the main pattern we want people to follow, I would recommend striking this and simply introducing the PUT LRO section later.

in [Create (PUT) with additional long-running processing](#create-put-with-additional-long-running-processing).
The remainder of this section describes the pattern for long-running POST and DELETE operations.

This diagram illustrates how a long-running operation with a status monitor is initiated and then how the client
determines it has completed and obtains its results:

```mermaid
sequenceDiagram
    participant Client
    participant API Endpoint
    participant Status Monitor
    Client->>API Endpoint: POST/DELETE

>     API Endpoint->>Client: HTTP/1.1 202 Accepted<br/>Retry-After: 5<br/>{ "id": "22", "status": "NotStarted" }

Reviewer 4: The `Retry-After` header here is a bit ambiguous. It could easily be interpreted as "retry the POST".

Author: We should leave Retry-After off of the response to the operation that initiates an LRO.  Then the client would initiate the get on the status monitor immediately, and the status monitor would then give the Retry-After in its response.

of the status monitor resource (without an api-version query parameter).

3. After waiting at least the amount of time specified by the previous response's `Retry-after` header,
the client issues a GET request to the status monitor using the ID in the body of the initial response.
The GET operation for the status monitor is documented in the REST API definition and the ID
is the last URL path segment.

4. The status monitor responds with information about the operation including its current status,
which should be represented as one of a fixed set of string values in a field named `status`.
If the operation is still being processed, the status field will contain a "non-terminal" value, like `NotStarted` or `Running`.

5. After the operation processing completes, a GET request to the status monitor returns the status monitor with a status field set to a terminal value -- `Succeeded`, `Failed`, or `Canceled` -- that indicates the result of the operation.
If the status is `Failed`, the status monitor resource contains an `error` field with a `code` and `message` that describes the failure.

> If the status is `Succeeded` and the LRO is an Action operation, the operation results will be returned in the `results` field of the status monitor.

Reviewer 4: ```suggestion
If the status is `Succeeded` and the LRO is an Action operation, the operation results will be returned in the `result` field of the status monitor.
```

If the status is `Succeeded` and the LRO is an operation on a resource, the client can perform a GET on the resource
to observe the result of the operation if desired.

6. There may be some cases where a long-running operation can be completed before the response to the initial request.

> In these cases, the operation should still return a `202 Accepted` with the `status` property set to the appropriate terminal state.

Reviewer: "and no retry-after header"

> 7. The service will auto-purge the status monitor resource after completion (at least 24 hours).

Reviewer 2: ```suggestion
7. The service is responsible for purging the status-monitor resource. It should auto-purge the status monitor resource after completion (at least 24 hours).
```

The service may offer DELETE of the status monitor resource due to GDPR/privacy.

### Long-running Action Operations

An action operation that is also long-running combines the [Action Operations](#action-operations) pattern
with the [Long Running Operations](#long-running-operations) pattern.

The operation is initiated with a POST operation and the operation path ends in `:action`.

```text

> POST /<service-or-resource-url>:action?api-version=2022-05-01

Reviewer 2: Are we suggesting that what goes at the end is the actual word *action*, instead of an "action verb"?

Reviewer: +1 for something like `<action>` or `<action-name>` if that's the templating style we're using everywhere

Operation-Id: 22 
 
{ 
   "arg1": 123 
   "arg2": "abc" 
} 
```

The response is a `202 Accepted` as described above.

```text
HTTP/1.1 202 Accepted
Operation-Location: https://<status-monitor-endpoint>/22
Retry-After: 5 
 
{
   "id": "22",
   "status": "NotStarted"
}
```

The client will issue a GET to the status monitor to obtain the status and result of the operation.

```text
GET https://<status-monitor-endpoint>/22?api-version=2022-05-01
```

When the operation completes successfully, the result (if there is one) will be included in the `result` field of the status monitor.

```text
HTTP/1.1 200 OK
 
{
   "id": "22",
   "status": "Succeeded",
   "result": { ... }
}
```

### PUT with additional long-running processing

> A special case of long-running operation that occurs often is a PUT operation to create or replace a resource

Reviewer 2: Re-reading this again,  it's not 100% clear when I would use this pattern. Or maybe said a slightly different way... If someone is reading this, how will they recognize that they have this pattern themselves? Maybe we can give an example from a couple of different domains. Some ideas...

"You may encounter this pattern when you have a one resource that either represents a physical resource. For example, in the Internet of Things domain, it is common to have a logical representation of a physical device. You can provision the logical device, but it is dependent on its real world counterpart to be installed, connected, and turned on. Similarly, you may have a resource that acts as a 'factory', creating other resources. An example of this could be a model used to train a cognitive service. The model itself needs to be created, but it will also produce results, e.g. a document with analyzed text.

that involves some additional long-running processing.
One example is a resource requires physical resources (e.g. servers) to be "provisioned" to make the resource functional.
In this case, the request may contain an `operation-id` header that the service will use as
the ID of the status monitor created for the operation.

> 

Reviewer 2: The text refers to `operation-id` but the example has `operation-Id`, capital I. We should be consistent.

If the resource supports ETags, the response may contain an `etag` header and possibly an `etag` property in the resource.

```text
HTTP/1.1 201 Created 
Operation-Id: 22
Operation-Location: https://items/operations/22
etag: "123abc"

{
  "id": "FooBar", 
  "etag": "123abc",
  "prop1": 555,
  "prop2": "something"
}
```

The client will issue a GET to the status monitor to obtain the status of the operation performing the additional processing.

```text
GET https://items/operations/22?api-version=2022-05-01
```

When the additional processing completes, the status monitor will indicate if it succeeded or failed.

```text
HTTP/1.1 200 OK
 
{
   "id": "22",
   "status": "Succeeded"
}
```

If the additional processing failed, the service may delete the original resource if it is not usable in this state.

### Long-running delete operation

A long-running delete operation follows the general pattern of a long-running operation --
it returns a `202 Accepted` with a status monitor which the client uses to determine the outcome of the delete.

The resource being deleted should remain visible (returned from a GET) until the delete operation completes successfully.

> When the delete operation completes successfully, a client must be able to create new resource with same name without conflicts.

Reviewer 2: The *must* in here feels a bit strong. If I was using a guid, must I really be able to create a new thing with an old guid?

### Controlling a long-running operation

It might be necessary to support some control action on a long-running operation, such as cancel.
This is implemented as a POST on the status monitor endpoint with `:action` added.

```text
POST /<status-monitor-url>:cancel?api-version=2022-05-01
```

A successful response to a control operation should be a `200 OK` with a representation of the status monitor.

```text
HTTP/1.1 200 OK 
 
{
   "id": "22",
   "status": "Canceled"
}

> ```

Reviewer: Do we want prescriptive guidance on the path structure for listing all active operations?  Either globally or per operation?  Or is that coming in a future update?
